schema/pgsql/schema.sql

```diff
@@ -344,89 +362,106 @@ CREATE TABLE rule_escalation (
     deleted boolenum NOT NULL DEFAULT 'n',
 
     CONSTRAINT pk_rule_escalation_recipient PRIMARY KEY (id),
-
-    CHECK (num_nonnulls(contact_id, contactgroup_id, schedule_id) = 1)
+    CONSTRAINT ck_rule_escalation_recipient_one_recipient CHECK (num_nonnulls(contact_id, contactgroup_id, schedule_id) = 1),
+    CONSTRAINT fk_rule_escalation_recipient_rule_escalation FOREIGN KEY (rule_escalation_id) REFERENCES rule_escalation(id),
+    CONSTRAINT fk_rule_escalation_recipient_contact FOREIGN KEY (contact_id) REFERENCES contact(id),
+    CONSTRAINT fk_rule_escalation_recipient_contactgroup FOREIGN KEY (contactgroup_id) REFERENCES contactgroup(id),
+    CONSTRAINT fk_rule_escalation_recipient_schedule FOREIGN KEY (schedule_id) REFERENCES schedule(id),
+    CONSTRAINT fk_rule_escalation_recipient_channel FOREIGN KEY (channel_id) REFERENCES channel(id)
 );
 
 CREATE INDEX idx_rule_escalation_recipient_changed_at ON rule_escalation_recipient(changed_at);
 
 CREATE TABLE incident (
     id bigserial,
-    object_id bytea NOT NULL REFERENCES object(id),
+    object_id bytea NOT NULL,
     started_at bigint NOT NULL,
     recovered_at bigint,
     severity severity NOT NULL,
 
-    CONSTRAINT pk_incident PRIMARY KEY (id)
+    CONSTRAINT pk_incident PRIMARY KEY (id),
+    CONSTRAINT fk_incident_object FOREIGN KEY (object_id) REFERENCES object(id)
 );
 
 CREATE TABLE incident_event (
-    incident_id bigint NOT NULL REFERENCES incident(id),
-    event_id bigint NOT NULL REFERENCES event(id),
+    incident_id bigint NOT NULL,
+    event_id bigint NOT NULL,
 
-    CONSTRAINT pk_incident_event PRIMARY KEY (incident_id, event_id)
+    CONSTRAINT pk_incident_event PRIMARY KEY (incident_id, event_id),
+    CONSTRAINT fk_incident_event_incident FOREIGN KEY (incident_id) REFERENCES incident(id),
+    CONSTRAINT fk_incident_event_event FOREIGN KEY (event_id) REFERENCES event(id)
 );
 
 CREATE TYPE incident_contact_role AS ENUM ('recipient', 'subscriber', 'manager');
 
 CREATE TABLE incident_contact (
-    incident_id bigint NOT NULL REFERENCES incident(id),
-    contact_id bigint REFERENCES contact(id),
-    contactgroup_id bigint REFERENCES contactgroup(id),
-    schedule_id bigint REFERENCES schedule(id),
+    incident_id bigint NOT NULL,
+    contact_id bigint,
+    contactgroup_id bigint,
+    schedule_id bigint,
     role incident_contact_role NOT NULL,
 
-    CONSTRAINT key_incident_contact_contact UNIQUE (incident_id, contact_id),
-    CONSTRAINT key_incident_contact_contactgroup UNIQUE (incident_id, contactgroup_id),
-    CONSTRAINT key_incident_contact_schedule UNIQUE (incident_id, schedule_id),
-    CONSTRAINT nonnulls_incident_recipients_check CHECK (num_nonnulls(contact_id, contactgroup_id, schedule_id) = 1)
+    CONSTRAINT uk_incident_contact_contact UNIQUE (incident_id, contact_id),
+    CONSTRAINT uk_incident_contact_contactgroup UNIQUE (incident_id, contactgroup_id),
+    CONSTRAINT uk_incident_contact_schedule UNIQUE (incident_id, schedule_id),
```

Comment:
These three constraints are referenced in the Go code and have to be changed there as well: https://github.com/Icinga/icinga-notifications/blob/860ea0ce3b74a0e8e463b3c2eccbe78fb7b58f62/internal/incident/db_types.go#L43-L54

Reply:
Maybe adding a comment in the schema to remind one about this would be helpful for possible future changes.

Reply:
Also, I assume you guys are aware of MySQL not having an equivalent. Hence, it will update whenever ANY constraint conflicts.

Reply:
You mean using a specific unique constraint in `ON CONFLICT`? We don't that because we need it, but because we [have to](https://www.postgresql.org/docs/current/sql-insert.html#SQL-ON-CONFLICT):

> For `ON CONFLICT DO UPDATE`, a *`conflict_target`* *must* be provided.

The code would be happy with an "on any conflict", we're just lucky to be able to predict which one would produce the conflict (due to the `num_nonnulls` check).

Reply:
Oh and regarding adding a comment: I think that totally makes sense and has nothing to do with expected further changes. It wouldn't have been a bad idea adding one right from the beginning, probably nobody was expecting it to change at that time either.

Reply:
> Which future changes? The only reason for _this_ constant renaming is the (OP) issue that they're not named at all. And even that took me lots of persuading.

Well, I don't know which future change will come. That's the thing about the future, right?

In the current situation, however, I would not have expected some constraint to be referred in the code. Thus, adding a one line comment might safe an hour of debugging in the future.